# Step 1. Open the data file and have a look at the general information.

In [1]:
import warnings
warnings.filterwarnings('ignore')
import pandas as pd
try:
    credit_scoring = pd.read_csv('C:/practicum/2. Data Preprocessing/7. Course Project/credit_scoring_eng.csv')
except:
    credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_scoring.info()
credit_scoring.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

## conclusion:

It seems that this is a large database, and in the columns **days_employed**, **total_income**  there are some amounts of missing values.
I will substitute the missing values with proper ones in the following step.

# Step 2. Data preprocessing

## Processing missing values

In [2]:
missing_values = credit_scoring.isnull().sum()
missing_values

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [3]:
print(credit_scoring['days_employed'].min())
print(credit_scoring['days_employed'].max())
print(credit_scoring['days_employed'].mean())
print(credit_scoring['days_employed'].median())

-18388.949900568383
401755.40047533
63046.497661473615
-1203.369528770489


In [4]:
print(credit_scoring['total_income'].min())
print(credit_scoring['total_income'].max())
print(credit_scoring['total_income'].mean())
print(credit_scoring['total_income'].median())

3306.762
362496.645
26787.56835465871
23202.87


In [5]:
credit_scoring['days_employed'] = credit_scoring['days_employed'].fillna(credit_scoring['days_employed'].median())
credit_scoring['total_income'] = credit_scoring['total_income'].fillna(credit_scoring['total_income'].median())
missing_values = credit_scoring.isnull().sum()
missing_values

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

## Conclusion:

1. We can see from the output that there are **2174 missing values** in each of these 2 columns: **days_employed** and **total_income**. And from step 1 we know that the datas in these 2 columns are both float64, which is quantitative.
2. In my opinion, total income and days employed are actually strongly connected variables because total income can be calculated from the days employed. *Perhaps people who are at a higher risk of default are likely to hide this information in order to get loan from the bank.*
3. Looking into the statistics of the 2 columns, we can see that in the column days_employed there is a large difference between the mean and median. It is probably because of some extreme values which significantly influences the mean. So I chose the median to represent the general values in this column. The difference between the median and mean in the column total_income is not that large, and I chose **median** to represent the values as well.
4. The missing values in the 2 columns are filled with the median of the existing values in each column. Now there are no missing values any more.

## Data type replacement

In [6]:
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [7]:
credit_scoring['days_employed'] = credit_scoring['days_employed'].astype('int64')
credit_scoring['total_income'] = credit_scoring['total_income'].astype('int64')
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


## Conclusion:

1. First check the information of the database again, we can see that only the days_employed and total_income columns are of the type **float64** and can be converted into **int64**.
2. Using the astype() method, I successfully changed these 2 columns into int64.

## Processing duplicates

In [8]:
print(credit_scoring['education'].unique())
print(credit_scoring['family_status'].unique())
print(credit_scoring['gender'].unique())
print(credit_scoring['income_type'].unique())
print(credit_scoring['purpose'].unique())

["bachelor's degree" 'secondary education' 'Secondary Education'
 'SECONDARY EDUCATION' "BACHELOR'S DEGREE" 'some college'
 'primary education' "Bachelor's Degree" 'SOME COLLEGE' 'Some College'
 'PRIMARY EDUCATION' 'Primary Education' 'Graduate Degree'
 'GRADUATE DEGREE' 'graduate degree']
['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried']
['F' 'M' 'XNA']
['employee' 'retiree' 'business' 'civil servant' 'unemployed'
 'entrepreneur' 'student' 'paternity / maternity leave']
['purchase of the house' 'car purchase' 'supplementary education'
 'to have a wedding' 'housing transactions' 'education' 'having a wedding'
 'purchase of the house for my family' 'buy real estate'
 'buy commercial real estate' 'buy residential real estate'
 'construction of own property' 'property' 'building a property'
 'buying a second-hand car' 'buying my own car'
 'transactions with commercial real estate' 'building a real estate'
 'housing' 'transactions with my real estate' 'cars' 'to beco

In [9]:
credit_scoring['education'] = credit_scoring['education'].str.lower()
print(credit_scoring['education'].unique())

["bachelor's degree" 'secondary education' 'some college'
 'primary education' 'graduate degree']


In [10]:
# now look into the duplicates inside the database
credit_scoring.duplicated().sum()

71

In [11]:
# drop the duplicates
credit_scoring = credit_scoring.drop_duplicates()
credit_scoring.duplicated().sum()

0

In [12]:
from nltk.stem import SnowballStemmer 
english_stemmer = SnowballStemmer('english')
for word in ['estate', 'property', 'house', 'housing', 'education', 'educated','university','car','wedding']:
    print(english_stemmer.stem(word), end = ' ')

estat properti hous hous educ educ univers car wed 

In [13]:
# To unit all the purposes of taking loans into 4 big catergories: real estate and property, car, education and wedding
def purpose_category(row):
    for word in row.split():
        word_stem = english_stemmer.stem(word) 
        if word_stem == 'estat' or word_stem == 'properti' or word_stem == 'hous':
            return 'real estate & property'
        if word_stem == 'car':
            return 'car'
        if word_stem == 'wed':
            return 'wedding'
        if word_stem == 'educ' or word_stem == 'univers':
            return 'education'
credit_scoring['purpose'] = credit_scoring['purpose'].apply(purpose_category)
print(credit_scoring['purpose'].unique())


['real estate & property' 'car' 'education' 'wedding']


In [14]:
credit_scoring.head()

children  days_employed  dob_years            education  education_id  \
0         1          -8437         42    bachelor's degree             0   
1         1          -4024         36  secondary education             1   
2         0          -5623         33  secondary education             1   
3         3          -4124         32  secondary education             1   
4         0         340266         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0         40620   
1            married                 0      F    employee     0         17932   
2            married                 0      M    employee     0         23341   
3            married                 0      M    employee     0         42820   
4  civil partnership                 1      F     retiree     0         25378   

                  purpose  
0  real estate & property  
1                     car  
2  real estate & property  
3               education  
4                 wedding

## Conclusion:

1. Check the unique values of each column which is not of the int64 type. From the unique values, we can see that in the education columns some values should be the same but appeared differently due to the uppercase of letter. In the purpose column there are many reasons that are actually of the same meaning but expressed differently. These duplicates should be dealt with.
2. Use the *str.lower()* method to change the strings in education column to lowercase and  use the *drop_duplicates()* method to drop the duplicates. Now we can see that there is no longer duplicate in this column.
3. Catergorize the different purposes into 4 main types: **real estate & property, car, education and wedding**. This is done by using *stemming*.

## Categorizing Data

In [15]:
'''some of the days_employed values are negative numbers, it looks strange. I want to look into the relationship between 
it and the income_type'''
employment_tab = credit_scoring[['income_type', 'days_employed']]
def check_days(days):
    if days >= 0:
        return 1
    else:
        return -1
employment_tab['check_days'] = credit_scoring['days_employed'].apply(check_days)
employment_tab.groupby(['check_days', 'income_type']).count()

days_employed
check_days income_type                               
-1         business                              5078
           civil servant                         1457
           employee                             11084
           entrepreneur                             2
           paternity / maternity leave              1
           retiree                                386
           student                                  1
 1         retiree                               3443
           unemployed                               2

In [16]:
'''some of the children values are negative numbers, it looks strange. I want to look into the relationship between 
it and the family_status,income_type '''
child_family = credit_scoring[['family_status', 'children']]
def check_child(children):
    if children >= 0:
        return 1
    else:
        return -1
child_family['check_child'] = credit_scoring['children'].apply(check_child)
child_family.groupby(['check_child','family_status']).count()

children
check_child family_status              
-1          civil partnership         5
            divorced                  4
            married                  29
            unmarried                 5
            widow / widower           4
 1          civil partnership      4146
            divorced               1191
            married               12310
            unmarried              2805
            widow / widower         955

## Conclusion: 

1. Looking at the days_employed and income_type values above, it seems that the **positive days_employed values** only lie in the 2 kinds of income_type: **retiree and unemployed**. I cannot find the logic why their days employed are negative while the other have negative values because of lack of addition information. *I guess this is some ways of catergory of the bank system.* 	
2. I also grouped the children and family_status values. It seems that the **negative children values** can be found in all kinds of families. It is also unclear why there are these negative values because of lack of addition information. *I think it is possible that the negative number of children might mean some unfortunate things like the children died at an early age, or it might be a miss-input of 1.*

# Step 3. Answer these questions

## Is there a relation between having kids and repaying a loan on time?

In [17]:
# look into the relationship between children and debt
child_debt = credit_scoring[['children', 'debt']]
# changing the negative chilren numbers by treating it as a miss-intput 1
child_debt.loc[child_debt['children'] == -1, 'children'] = 1
child_debt.groupby('children').mean()

debt
children          
0         0.075438
1         0.091658
2         0.094542
3         0.081818
4         0.097561
5         0.000000
20        0.105263

## Conclusion: 

It seems that the family with the most children (20) are likely to have a highest average level of debt. But the level of debt doesn't actually grow as the number of children increase. *So I would say that the level of debt might be positively influenced by the number of children, but it can also result from other factors.* 

## Is there a relation between marital status and repaying a loan on time? 

In [18]:
# look into the relationship between family_status and debt
family_debt = credit_scoring[['family_status', 'debt']]
family_debt.groupby('family_status').mean()

debt
family_status              
civil partnership  0.093471
divorced           0.071130
married            0.075452
unmarried          0.097509
widow / widower    0.065693

## Conclusion: 

From the data above, we can see that people with a family status of **'civil partnership'** and **'unmarried'** are *most likely to bear debt*, while people of **widor or widower** are *least likely to bear debt*.

## Is there a relation between income level and repaying a loan on time? 

In [19]:
# look into the relationship between total_income and debt
income_debt = credit_scoring[['debt']]
def income_level(income):
    if income >= 100000:
        return 'very high'
    elif income >= 50000:
        return 'high'
    elif income >= 20000:
        return 'middle'
    elif income >= 5000:
        return 'low'
    else:
        return 'very low'
income_debt['income_level'] = credit_scoring['total_income'].apply(income_level)
income_debt.groupby('income_level').mean()

debt
income_level          
high          0.070434
low           0.082528
middle        0.081551
very high     0.060606
very low      0.076923

## Conclusion: 

1. I used a new function income_level to redefine the total income levels of these people, from **very high (>= 100000), high (>= 50000), middle (>= 20000), low (>= 5000) to very low (< 5000)**. *(This classfication is only a rough idea of mine, it might be different from those in reality.)*
2. The group with highest income level are least likely to be in debt. *From the low level to very high level, the average debt decreases. But the very low group don't actually bear the most debt. Perhaps it is because they know that they have so little to pay back the debt, so they would like to save their money instead of borrowing more. Another reason might be that financial institutions like banks are generally less willing to lend money to them because they are less likely to pay the loans back.*

##  How do different loan purposes affect on-time repayment of the loan? 

In [20]:
# look into the relationship between purpose and debt
purpose_debt = credit_scoring[['purpose', 'debt']]
purpose_debt.groupby('purpose').mean()

debt
purpose                         
car                     0.093590
education               0.092200
real estate & property  0.072334
wedding                 0.080034

## Conclusion: 

From the data above, we can see that people are *most likely* into debt for buying a **car**, then for education and wedding. They are *least likely* to take loans for costs related to **real estate or property**.

# Step 4. General conclusion

We can conclude from the above analysis that children, marital status, income level and loan purposes all have influence on the debt people are in. While there is *a siginificant negative relationship between income level and debt level (except for the very low income group)*, the *relationship between the number of children and debt level isn't that clear*. People who are of the **unmarried** family status and people who want to buy a **car** are *most likely* to take loans from banks.